src/main/java/beam/physsim/cch/RoutingFrameworkTravelTimeCalculator.scala

```diff
@@ -0,0 +1,230 @@
+      Math.max(Runtime.getRuntime.availableProcessors() / 4, 1),
+      new ThreadFactoryBuilder().setDaemon(true).setNameFormat("routing-framework-worker-%d").build()
+    )
+  )
+
+  private val odsFactor: Int =
+    Math.max(
+      (1.0 / beamServices.beamConfig.beam.agentsim.agentSampleSizeAsFractionOfPopulation *
+      beamServices.beamConfig.beam.physsim.cch.congestionFactor).toInt,
+      1
+    )
+
+  private val graph: RoutingFrameworkGraph = routingFrameworkWrapper.generateGraph()
+
+  private val coord2VertexId: Map[Coordinate, Long] =
+    graph.vertices.map(x => x.coordinate -> x.id).toMap
+
+  private val coordKey = (coord: Coordinate) => (coord.x * 10).toInt + "|" + (coord.y * 10).toInt
+
+  private val coordKey2Coordinates: Map[String, Seq[Vertex]] = graph.vertices
+    .map(v => coordKey(v.coordinate) -> v)
+    .groupBy(_._1)
+    .mapValues(_.map(_._2))
+
+  def getLink2TravelTimes(
+    pathTraversalEvents: util.Collection[PathTraversalEvent],
+    iterationEndsEvent: IterationEndsEvent,
+    links: util.Collection[_ <: Link],
+    maxHour: Int
+  ): java.util.Map[String, Array[Double]] = {
+    val iterationNumber: Int = iterationEndsEvent.getIteration
+    val startTime: Long = System.currentTimeMillis
+
+    logger.info("Finished creation of graph {}", System.currentTimeMillis - startTime)
+
+    val id2Link = links.toStream.map(x => x.getId.toString.toInt -> x).toMap
+
+    val hour2Events: Map[Int, List[TravelInfo]] = generateHour2Events(pathTraversalEvents)
+
+    val odsCreationFutures = hour2Events
+      .map {
+        case (hour, events) =>
+          Future {
+            val stopWatch: StopWatch = new StopWatch
+            stopWatch.start()
+
+            var odNumber = 0
+
+            val ods = events.toStream
+              .map { event =>
+                for {
+                  firstLinkId <- event.linkIds.headOption
+                  lastLinkId  <- event.linkIds.lastOption
+                  firstWayId  <- linkWayId(id2Link(firstLinkId))
+                  secondWayId <- linkWayId(id2Link(lastLinkId))
+                  origin      <- osmInfoHolder.getCoordinatesForWayId(firstWayId).headOption
+                  destination <- osmInfoHolder.getCoordinatesForWayId(secondWayId).lastOption
+                } yield origin -> destination
+              }
+              .collect { case Some(od @ (origin, destination)) if origin != destination => od }
+              .map {
+                case (origin, destination) =>
+                  val firstVertexId: Long = coord2VertexId
+                    .getOrElse(origin, getClosestVertexId(origin))
+                  val secondVertexId: Long = coord2VertexId
+                    .getOrElse(destination, getClosestVertexId(destination))
+
+                  odNumber = odNumber + 1
+
+                  OD(firstVertexId, secondVertexId)
+              }
+
+            val odStream = ods
+              .flatMap(od => Stream.range(0, odsFactor, 1).map(_ => od))
```

Comment:
If I'm reading this correctly it could be:
```suggestion
            val odStream = ods.flatMap(Seq.fill(odsFactor)(_))
```

Reply:
Will it still be lazy?